### DISCUSSION TITLE
Get ECSQL used by Presentation Library

### DISCUSSION THREAD

**Discussion by benjamin-lindsey** [iTwin/itwinjs-core#5999] [Q&A] [answered]
Hi All,

[It was recently suggested](https://github.com/iTwin/itwinjs-core/discussions/5981) that I check out the Presentation library for easily extracting properties without having to know specifics about item types. This worked out great. 

Now I'm curious if there's a way to get the same ECSQL that the Presentation library is using. [From the docs:](https://www.itwinjs.org/presentation/motivation/)

"It's built on top of ECSQL to help developers easily get data from iModels in a format that's ready to be displayed to end users."

Where is this ECSQL? Is it available to us?

Thanks!
Ben

**Comment by pmconne** [comment]
> Where is this ECSQL? Is it available to us?

It is [pretty complicated](https://github.com/iTwin/imodel-native/tree/main/iModelCore/ECPresentation) to produce.

**Reply by benjamin-lindsey** [reply]
What about the resultant ECSQL statement(s) that the Presentation library generates? For example, if `getElementProperties` generates ECSQL behind-the-scenes, is it obtainable?

Seems like there might need to be code between several queries which is where the complexity comes in. Of course I have no idea.

**Reply by grigasp** [reply]
Specific ECSQL queries used by Presentation library are not available and even if they were, they wouldn't be usable externally due to use of various [custom SQLite functions](https://www.sqlite.org/appfunc.html) that are only available when the queries are executed in the context of Presentation library.

However, you can get metadata about the properties for specific element(s), which allows you to build a similar query. The metadata can be retrieved in a form of a [content Descriptor](https://www.itwinjs.org/reference/presentation-common/content/descriptor/) by calling [PresentationManager.getContentDescriptor](https://www.itwinjs.org/reference/presentation-frontend/core/presentationmanager/#getcontentdescriptor). You'll need to pass a [Ruleset](https://www.itwinjs.org/reference/presentation-common/presentationrules/ruleset/) and a [KeySet](https://www.itwinjs.org/reference/presentation-common/core/keyset/):
- For ruleset, I recommend just using the [DEFAULT_PROPERTY_GRID_RULESET](https://www.itwinjs.org/reference/presentation-components/propertygrid/default_property_grid_ruleset/).
- For `KeySet`, just create an empty one and add element's `InstanceKey` in there.

The descriptor contains information about the class that you should select from and the list of all properties that should be selected along with necessary relationships' information to create joins to other classes.

**Reply by benjamin-lindsey** [reply]
@grigasp - It seems [InstanceKey](https://www.itwinjs.org/reference/presentation-common/core/instancekey/) requires a className. Does this approach assume we know the className, whether it's obtained by getElementProperties or by knowledge of the element's item type schema?

**Reply by grigasp** [reply, marked as answer]
If requesting content for some bis.Element, you can pass "BisCore:Element" as the class name.

**Reply by benjamin-lindsey** [reply]
Perfect, that did the trick! Thanks for your help.
